src/zarr/codecs/pipeline/hybrid.py

```diff
@@ -0,0 +1,115 @@
+from __future__ import annotations
+
+from itertools import islice
+from typing import TYPE_CHECKING, TypeVar
+import numpy as np
+from dataclasses import dataclass
+
+from zarr.config import config
+from zarr.abc.codec import (
+    ByteGetter,
+    ByteSetter,
+    Codec,
+)
+from zarr.codecs.pipeline.batched import BatchedCodecPipeline
+from zarr.codecs.pipeline.core import CodecPipeline
+from zarr.common import concurrent_map
+
+if TYPE_CHECKING:
+    from typing import List, Optional, Tuple, Iterable
+    from typing_extensions import Self
+    from zarr.common import ArraySpec, BytesLike, SliceSelection
+
+T = TypeVar("T")
+
+
+def batched(iterable: Iterable[T], n: int) -> Iterable[Tuple[T, ...]]:
+    if n < 1:
+        raise ValueError("n must be at least one")
+    it = iter(iterable)
+    while batch := tuple(islice(it, n)):
+        yield batch
+
+
+@dataclass(frozen=True)
+class HybridCodecPipeline(CodecPipeline):
```

Comment:
What are the operational differences between this class and the `BatchedCodecPipeline`?

Reply:
This is a good question. `BatchedCodecPipeline` executes all steps in lock step for all chunks that are requested. `HybridCodecPipeline` adds a batch size where each batch can run asynchronously interleaved. We could consolidate both into 1 class. That would leave the abstract `CodecPipeline` (for implementing custom codec pipelines) and one implementation.